flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateLocalitySlotAssigner.java

```diff
@@ -0,0 +1,209 @@
+                                        evi ->
+                                                parallelism.merge(
+                                                        evi.getJobVertexId(), 1, Integer::sum)));
+
+        PriorityQueue<AllocationScore> scores = new PriorityQueue<>(Comparator.reverseOrder());
+        for (ExecutionSlotSharingGroup group : groups) {
+            calculateScore(group, parallelism)
+                    .forEach(
+                            (allocationId, score) ->
+                                    scores.add(
+                                            new AllocationScore(
+                                                    group.getId(), allocationId, score)));
+        }
+        Map<String, ExecutionSlotSharingGroup> groupsById =
+                groups.stream().collect(toMap(ExecutionSlotSharingGroup::getId, identity()));
+        Map<AllocationID, SlotInfo> slotsById =
+                slots.stream().collect(toMap(SlotInfo::getAllocationId, identity()));
+        List<ExecutionSlotSharingGroupAndSlot> result = new ArrayList<>();
+        AllocationScore score;
+        while ((score = scores.poll()) != null) {
+            SlotInfo slot = slotsById.remove(score.getAllocationId());
+            if (slot != null) {
+                ExecutionSlotSharingGroup group = groupsById.remove(score.getGroup());
+                if (group != null) {
+                    result.add(new ExecutionSlotSharingGroupAndSlot(group, slot));
+                }
+            }
+        }
+
+        // Distribute the remaining slots with no score
+        Iterator<? extends SlotInfo> remainingSlots = slotsById.values().iterator();
+        for (ExecutionSlotSharingGroup group : groupsById.values()) {
+            result.add(new ExecutionSlotSharingGroupAndSlot(group, remainingSlots.next()));
+            remainingSlots.remove();
+        }
+
+        return AssignmentResult.of(result, remainingSlots);
+    }
+
+    public Map<AllocationID, Integer> calculateScore(
+            ExecutionSlotSharingGroup group, Map<JobVertexID, Integer> parallelism) {
+        final Map<AllocationID, Integer> score = new HashMap<>();
+        for (ExecutionVertexID evi : group.getContainedExecutionVertices()) {
+            if (maxParallelism.containsKey(evi.getJobVertexId())) {
+                final KeyGroupRange kgr =
+                        KeyGroupRangeAssignment.computeKeyGroupRangeForOperatorIndex(
+                                maxParallelism.get(evi.getJobVertexId()),
+                                parallelism.get(evi.getJobVertexId()),
+                                evi.getSubtaskIndex());
+                locality.forEach(
+                        (allocationId, potentials) -> {
+                            KeyGroupRange prev = potentials.get(evi.getJobVertexId());
+                            if (prev != null) {
+                                int intersection = prev.getIntersection(kgr).getNumberOfKeyGroups();
```

Comment:
I agree, that would make sense. And that's also something doable - the state size can be obtained from `ExecutionGraph.CheckpointCoordinator.CompletedCheckpointStore.LatestCheckpoint`.
I was just thinking to make it a followup, but it can also be implemented in this PR.

Reply:
I've added state size estimation in dbdec757acc3ea362d375d4231779cffe400dbb5. My intention is to validate the interfaces. Before merging this PR, I'd drop the commit and move it to a separate ticket/PR, just to make this one smaller.